File: oop/python/player.py
"""Concrete strategies: RandomPlayer + OptimalPlayer.

Python idiom notes:
    - The Strategy pattern here is almost invisible: a class with one
      method. In smaller Python code you might just use a function/lambda;
      we keep the class form to mirror the other languages.
    - OptimalPlayer is generic in the same sense Game is: it never names
      a concrete game type. It calls legal_moves / apply_move / undo_move /
      is_over / winner / current_side / state_key. The whole search lives
      on the Game contract.
"""

from __future__ import annotations

import random

from game import Game, Player, Side


class RandomPlayer(Player):
    def __init__(self, seed: int, label: str = "Random"):
        # A private RNG instance keeps this player's randomness independent
        # of other players and of the global random module.
        self._rng = random.Random(seed)
        self._label = label

    def choose_move(self, game: Game):
        return self._rng.choice(list(game.legal_moves()))

    @property
    def name(self) -> str:
        return self._label


class OptimalPlayer(Player):
    """Negamax with explicit transposition table.

    Returns from search are (outcome, depth) pairs from the *side to move*'s
    perspective:
      - outcome ∈ {-1, 0, +1}  (loss, draw, win under optimal play)
      - depth = plies-to-terminal under optimal play

    Move selection is fully deterministic:
      1. Maximize outcome.
      2. On a tied outcome, minimize depth when winning (finish fast)
         and maximize depth when tying or losing (drag it out).
      3. On full ties, take the first move legal_moves() yielded.

    The cache is keyed on game.state_key() — a compact integer per game,
    not a render string, so no formatting cost in the search loop.
    """

    def __init__(self, label: str = "Optimal"):
        self._label = label
        self._cache: dict[int, tuple[int, int]] = {}

    @property
    def name(self) -> str:
        return self._label

    def choose_move(self, game: Game):
        best_value: tuple[int, int] | None = None
        best_move = None
        for move in game.legal_moves():
            undo = game.apply_move(move)
            opp_outcome, opp_depth = self._search(game)
            game.undo_move(undo)
            value = (-opp_outcome, opp_depth + 1)
            if best_value is None or self._better(value, best_value):
                best_value = value
                best_move = move
        return best_move

    def _search(self, game: Game) -> tuple[int, int]:
        if game.is_over():
            w = game.winner()
            if w is None:
                return (0, 0)
            # Our games flip current_side at the end of apply_move, so when
            # the game is over with a winner, the side to move is the loser.
            return (-1, 0)

        key = game.state_key()
        cached = self._cache.get(key)
        if cached is not None:
            return cached

        best: tuple[int, int] | None = None
        for move in game.legal_moves():
            undo = game.apply_move(move)
            opp_outcome, opp_depth = self._search(game)
            game.undo_move(undo)
            value = (-opp_outcome, opp_depth + 1)
            if best is None or self._better(value, best):
                best = value

        # `best` is non-None: legal_moves was non-empty (game not over).
        self._cache[key] = best  # type: ignore[assignment]
        return best  # type: ignore[return-value]

    @staticmethod
    def _better(a: tuple[int, int], b: tuple[int, int]) -> bool:
        # Higher outcome wins. On ties: shorter depth when winning,
        # longer depth when tying or losing. Strict — equal returns False
        # so the first-encountered candidate is kept (determinism).
        if a[0] != b[0]:
            return a[0] > b[0]
        if a[0] > 0:
            return a[1] < b[1]
        return a[1] > b[1]

File: oop/rust/src/player.rs
// Two strategies: RandomPlayer + OptimalPlayer.
//
// RandomPlayer is the obvious one — uniform over legal moves.
// OptimalPlayer is a deterministic, memoized negamax. The whole search
// only ever calls trait methods on Game, so it works for any concrete
// game implementing the trait.

use std::collections::HashMap;

use crate::game::{Game, Player};
use crate::rng::XorShift64;

pub struct RandomPlayer {
    rng: XorShift64,
    label: String,
}

impl RandomPlayer {
    pub fn new(seed: u64, label: &str) -> Self {
        Self { rng: XorShift64::new(seed), label: label.to_string() }
    }
}

// `impl<M: Clone, U> Player<M, U> for RandomPlayer` says "I don't care
// what moves or undos look like, I just pick a move and never apply it."
// The Undo parameter is part of the Player trait signature even though
// RandomPlayer never uses it.
impl<M: Clone, U> Player<M, U> for RandomPlayer {
    fn choose_move(&mut self, game: &mut dyn Game<Move = M, Undo = U>) -> M {
        let moves = game.legal_moves();
        let i = self.rng.range(moves.len());
        moves[i].clone()
    }
    fn name(&self) -> &str {
        &self.label
    }
}

/// Deterministic negamax with explicit transposition table.
///
/// Search returns `(outcome, depth)` from the side-to-move's perspective:
///   outcome ∈ {-1, 0, +1}  (loss, draw, win under optimal play)
///   depth   = plies-to-terminal under optimal play
///
/// Move selection:
///   1. Maximize outcome.
///   2. On a tied outcome, minimize depth when winning (finish fast)
///      and maximize depth when tying or losing (drag it out).
///   3. On full ties, take the first move legal_moves() yielded.
///
/// The cache is keyed on `game.state_key()` — a u64, not a render
/// string — so the search loop allocates only what apply/undo cost.
pub struct OptimalPlayer {
    label: String,
    cache: HashMap<u64, (i8, u32)>,
}

impl OptimalPlayer {
    pub fn new(label: &str) -> Self {
        Self { label: label.to_string(), cache: HashMap::new() }
    }

    pub fn cache_size(&self) -> usize { self.cache.len() }

    fn better(a: (i8, u32), b: (i8, u32)) -> bool {
        if a.0 != b.0 { return a.0 > b.0; }
        if a.0 > 0 { a.1 < b.1 } else { a.1 > b.1 }
    }

    fn search<M: Clone, U>(&mut self, game: &mut dyn Game<Move = M, Undo = U>) -> (i8, u32) {
        if game.is_over() {
            return match game.winner() {
                None => (0, 0),
                // Our games flip current_side at the end of apply_move,
                // so when the game is over with a winner, the side to
                // move is the loser.
                Some(_) => (-1, 0),
            };
        }

        let key = game.state_key();
        if let Some(&v) = self.cache.get(&key) {
            return v;
        }

        let moves = game.legal_moves();
        let mut best: Option<(i8, u32)> = None;
        for m in &moves {
            let undo = game.apply_move(m);
            let (oo, od) = self.search(game);
            game.undo_move(undo);
            let value = (-oo, od + 1);
            if best.is_none() || Self::better(value, best.unwrap()) {
                best = Some(value);
            }
        }

        let v = best.unwrap();  // legal_moves was non-empty (game not over)
        self.cache.insert(key, v);
        v
    }
}

impl<M: Clone, U> Player<M, U> for OptimalPlayer {
    fn choose_move(&mut self, game: &mut dyn Game<Move = M, Undo = U>) -> M {
        let moves = game.legal_moves();
        let mut best: Option<(i8, u32)> = None;
        let mut best_move: Option<M> = None;
        for m in &moves {
            let undo = game.apply_move(m);
            let (oo, od) = self.search(game);
            game.undo_move(undo);
            let value = (-oo, od + 1);
            if best.is_none() || Self::better(value, best.unwrap()) {
                best = Some(value);
                best_move = Some(m.clone());
            }
        }
        best_move.unwrap()
    }
    fn name(&self) -> &str {
        &self.label
    }
}
